File: Matlab/Oct_28.m
%% Linear model extended version -- quadratic
n = 100;
X = 5*rand(n, 1);
Y = 3 + 2*X + 4*X.^2 + normrnd(0, 1, n, 1);
plot(X, Y, '.');

% Fitting linear model manually
X_des = [ones(size(X, 1), 1), X];
b_hat = (X_des'*X_des)\X_des'*Y;
Y_hat = X_des*b_hat;
figure;
plot(X, Y, '.'); hold on; plot(X, Y_hat, '.'); hold off;
legend('true values', 'estimated values');

% Use extended version
X_des = [ones(size(X, 1), 1), X, X.^2];
b_hat = (X_des'*X_des)\X_des'*Y;
Y_hat = X_des*b_hat;
figure;
plot(X, Y, '.'); hold on; plot(X, Y_hat, '.'); hold off;
legend('true values', 'estimated values');

%% Model Selection
%Generating Data 
n = 100;
X = 2*pi*(rand(n, 1)-1/2);
Y = 1 + 2*X + 4*sin(X) -2*cos(X)+normrnd(0, 1, n, 1);
plot(X, Y, '.');

%% Fitting linear model
fprintf("Model: Y ~ 1 + X\n")
X_des = [ones(size(X, 1), 1), X];
b_hat = (X_des'*X_des)\X_des'*Y;
Y_hat = X_des*b_hat;
figure;
plot(X, Y, '.'); hold on; plot(X, Y_hat, '.'); hold off;
legend('true values', 'estimated values');
title('Y ~ 1 + X')
% Computing leave-one-out cross validation error
cross_val_1 = 0;
for i = 1:n
    X_des_new = X_des;
    X_des_new(i, :) = [];
    Y_new = Y;
    Y_new(i) = [];
    b_hat = (X_des_new'*X_des_new)\X_des_new'*Y_new;
    y_hat = X_des(i, :) * b_hat;
    cross_val_1 = cross_val_1 + (Y(i) - y_hat)^2;
end
fprintf("RSS = %.3f\n", sum((Y-Y_hat).^2))
fprintf("AIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1))
fprintf("BIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1)/2*log(n))
fprintf("Leave-one-out error = %.3f\n", cross_val_1)

%% Fitting extended model with cos(X)
fprintf("Model: Y ~ 1 +  X + cos(X)\n")
X_des = [ones(size(X, 1), 1), X, cos(X)];
b_hat = (X_des'*X_des)\X_des'*Y;
Y_hat = X_des*b_hat;
figure;
plot(X, Y, '.'); hold on; plot(X, Y_hat, '.'); hold off;
legend('true values', 'estimated values');
title('Y ~ 1 + X + cos(X)')
% Computing leave-one-out cross validation error
cross_val_1 = 0;
for i = 1:n
    X_des_new = X_des;
    X_des_new(i, :) = [];
    Y_new = Y;
    Y_new(i) = [];
    b_hat = (X_des_new'*X_des_new)\X_des_new'*Y_new;
    y_hat = X_des(i, :) * b_hat;
    cross_val_1 = cross_val_1 + (Y(i) - y_hat)^2;
end
fprintf("RSS = %.3f\n", sum((Y-Y_hat).^2))
fprintf("AIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1))
fprintf("BIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1)/2*log(n))
fprintf("Leave-one-out error = %.3f\n", cross_val_1)

%% Fitting extended model with sin(X)
fprintf("Model: Y ~ 1 + X + sin(X)\n")
X_des = [ones(size(X, 1), 1), X, sin(X)];
b_hat = (X_des'*X_des)\X_des'*Y;
Y_hat = X_des*b_hat;
figure;
plot(X, Y, '.'); hold on; plot(X, Y_hat, '.'); hold off;
legend('true values', 'estimated values');
title('Y ~ 1 + X + sin(X)')
% Computing leave-one-out cross validation error
cross_val_1 = 0;
for i = 1:n
    X_des_new = X_des;
    X_des_new(i, :) = [];
    Y_new = Y;
    Y_new(i) = [];
    b_hat = (X_des_new'*X_des_new)\X_des_new'*Y_new;
    y_hat = X_des(i, :) * b_hat;
    cross_val_1 = cross_val_1 + (Y(i) - y_hat)^2;
end
fprintf("RSS = %.3f\n", sum((Y-Y_hat).^2))
fprintf("AIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1))
fprintf("BIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1)/2*log(n))
fprintf("Leave-one-out error = %.3f\n", cross_val_1)

%% Fitting extended model with both sin(X) and cos(X)
fprintf("Model: Y ~ 1 + X + sin(X) + cos(X)\n")
X_des = [ones(size(X, 1), 1), X, sin(X), cos(X)];
b_hat = (X_des'*X_des)\X_des'*Y;
Y_hat = X_des*b_hat;
figure;
plot(X, Y, '.'); hold on; plot(X, Y_hat, '.'); hold off;
legend('true values', 'estimated values');
title('Y ~ 1 + X + sin(X) + cos(X)')
% Computing leave-one-out cross validation error
cross_val_1 = 0;
for i = 1:n
    X_des_new = X_des;
    X_des_new(i, :) = [];
    Y_new = Y;
    Y_new(i) = [];
    b_hat = (X_des_new'*X_des_new)\X_des_new'*Y_new;
    y_hat = X_des(i, :) * b_hat;
    cross_val_1 = cross_val_1 + (Y(i) - y_hat)^2;
end
fprintf("RSS = %.3f\n", sum((Y-Y_hat).^2))
fprintf("AIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1))
fprintf("BIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1)/2*log(n))
fprintf("Leave-one-out error = %.3f\n", cross_val_1)

%% Fitting extended model with both sin(X) and cos(X) and X^2
fprintf("Model: Y ~ 1 + X + X^2 + sin(X) + cos(X)\n")
X_des = [ones(size(X, 1), 1), X, X.^2, sin(X), cos(X)];
b_hat = (X_des'*X_des)\X_des'*Y;
Y_hat = X_des*b_hat;
figure;
plot(X, Y, '.'); hold on; plot(X, Y_hat, '.'); hold off;
legend('true values', 'estimated values');
title('Y ~ 1 + X + X^2 + sin(X) + cos(X)')
% Computing leave-one-out cross validation error
cross_val_1 = 0;
for i = 1:n
    X_des_new = X_des;
    X_des_new(i, :) = [];
    Y_new = Y;
    Y_new(i) = [];
    b_hat = (X_des_new'*X_des_new)\X_des_new'*Y_new;
    y_hat = X_des(i, :) * b_hat;
    cross_val_1 = cross_val_1 + (Y(i) - y_hat)^2;
end
fprintf("RSS = %.3f\n", sum((Y-Y_hat).^2))
fprintf("AIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1))
fprintf("BIC = %.3f\n", -sum((Y-Y_hat).^2)-size(b_hat, 1)/2*log(n))
fprintf("Leave-one-out error = %.3f\n", cross_val_1)

% forward stepwise regression will choose models as follows:
% Y ~ 1
% Y ~ 1 + X
% Y ~ 1 + X + sin(X)
% Y ~ 1 + X + sin(X) + cos(X)
% Stop

% backward stepwise regression will choose models as follows:
% Y ~ 1 + X + X^2 + sin(X) + cos(X) 
% Y ~ 1 + X + sin(X) + cos(X)
% Stop

File: R/Oct_28.R
#### Linear model extended version - quadratic
n <- 100
X <- 5*matrix(runif(n), n, 1, byrow=TRUE)
Y <- 3 + 2*X + 4*X^2 + matrix(rnorm(n, 0, 2), n, 1, byrow=TRUE)
plot(X, Y, cex=.4)

# Fitting linear model manually
X_des <- cbind(matrix(rep(1, n), n, 1, byrow=TRUE), X)
b_hat <- solve(t(X_des) %*% X_des) %*% t(X_des) %*% Y
Y_hat <- X_des %*% b_hat
plot(X, Y, cex=.4, ylim=range(-20, 120),)
par(new=TRUE)
plot(X, Y_hat, cex=.4, col='red', ylim=range(-20, 120))

# Use extended version
X_des <- cbind(matrix(rep(1, n), n, 1, byrow=TRUE), X, X^2)
b_hat <- solve(t(X_des) %*% X_des) %*% t(X_des) %*% Y
Y_hat <- X_des %*% b_hat
plot(X, Y, cex=.4, ylim=range(-20, 120), ylab="")
par(new=TRUE)
plot(X, Y_hat, cex=.4, col='red', ylim=range(-20, 120))

#### Model Selection
# Generating Data
n <- 100
X <- 2*pi*matrix(runif(n)-1/2, n, 1, byrow=TRUE)
Y <- 1 + 2*X + 4*sin(X) - 2*cos(X) + matrix(rnorm(n, 0, 1), n, 1, byrow=TRUE)
plot(X, Y, cex=.4)

Performance <- data.frame(matrix(NA, nrow = 4, ncol = 5))
colnames(Performance) = c("1+X", "1+X+cos(X)", "1+X+sin(X)", "1+X+sin(X)+cos(X)", "1+X+X^2+sin(X)+cos(X)")
rownames(Performance) = c("RSS", "AIC", "BIC", "Cross-valid")

# Fitting linear model
X_des <- cbind(matrix(rep(1, n), n, 1, byrow=TRUE), X)
b_hat <- solve(t(X_des) %*% X_des) %*% t(X_des) %*% Y
Y_hat <- X_des %*% b_hat
plot(X, Y, cex=.4, ylab="", ylim=range(-12, 13))
par(new=TRUE)
plot(X, Y_hat, cex=.4, ylab="Y", ylim=range(-12, 13), title("Y ~ 1 + X"), col='red')
# Computing leave-one-out cross validation error
cross_val_1 <- 0
for (i in 1:n){
  X_des_new <- X_des[-i,]
  Y_new <- Y[-i]
  b_hat <- solve(t(X_des_new) %*% X_des_new) %*% t(X_des_new) %*% Y_new
  y_hat <- X_des[i,] %*% b_hat
  #print(cross_val_1)
  print((Y[i] - y_hat)^2)
  cross_val_1 = cross_val_1 + (Y[i] - y_hat)^2
  #print(length(cross_val_1))
}
RSS <- sum((Y - Y_hat)^2)
AIC <- -sum((Y - Y_hat)^2) - length(b_hat)
BIC <- -sum((Y - Y_hat)^2) - length(b_hat)/2*log(n)
Performance[,1] = c(RSS, AIC, BIC, cross_val_1)

# Fitting extended model with cos(X)
X_des <- cbind(matrix(rep(1, n), n, 1, byrow=TRUE), X, cos(X))
b_hat <- solve(t(X_des) %*% X_des) %*% t(X_des) %*% Y
Y_hat <- X_des %*% b_hat
plot(X, Y, cex=.4, ylab="", ylim=range(-12, 13))
par(new=TRUE)
plot(X, Y_hat, cex=.4, ylab="Y", ylim=range(-12, 13), title("Y ~ 1 + X"), col='red')
# Computing leave-one-out cross validation error
cross_val_1 <- 0
for (i in 1:n){
  X_des_new <- X_des[-i,]
  Y_new <- Y[-i]
  b_hat <- solve(t(X_des_new) %*% X_des_new) %*% t(X_des_new) %*% Y_new
  y_hat <- X_des[i,] %*% b_hat
  #print(cross_val_1)
  print((Y[i] - y_hat)^2)
  cross_val_1 = cross_val_1 + (Y[i] - y_hat)^2
  #print(length(cross_val_1))
}
RSS <- sum((Y - Y_hat)^2)
AIC <- -sum((Y - Y_hat)^2) - length(b_hat)
BIC <- -sum((Y - Y_hat)^2) - length(b_hat)/2*log(n)
Performance[,2] = c(RSS, AIC, BIC, cross_val_1)

# Fitting extended model with sin(X)
X_des <- cbind(matrix(rep(1, n), n, 1, byrow=TRUE), X, sin(X))
b_hat <- solve(t(X_des) %*% X_des) %*% t(X_des) %*% Y
Y_hat <- X_des %*% b_hat
plot(X, Y, cex=.4, ylab="", ylim=range(-12, 13))
par(new=TRUE)
plot(X, Y_hat, cex=.4, ylab="Y", ylim=range(-12, 13), title("Y ~ 1 + X"), col='red')
# Computing leave-one-out cross validation error
cross_val_1 <- 0
for (i in 1:n){
  X_des_new <- X_des[-i,]
  Y_new <- Y[-i]
  b_hat <- solve(t(X_des_new) %*% X_des_new) %*% t(X_des_new) %*% Y_new
  y_hat <- X_des[i,] %*% b_hat
  #print(cross_val_1)
  print((Y[i] - y_hat)^2)
  cross_val_1 = cross_val_1 + (Y[i] - y_hat)^2
  #print(length(cross_val_1))
}
RSS <- sum((Y - Y_hat)^2)
AIC <- -sum((Y - Y_hat)^2) - length(b_hat)
BIC <- -sum((Y - Y_hat)^2) - length(b_hat)/2*log(n)
Performance[,3] = c(RSS, AIC, BIC, cross_val_1)

# Fitting extended model with cos(X) and sin(X)
X_des <- cbind(matrix(rep(1, n), n, 1, byrow=TRUE), X, cos(X), sin(X))
b_hat <- solve(t(X_des) %*% X_des) %*% t(X_des) %*% Y
Y_hat <- X_des %*% b_hat
plot(X, Y, cex=.4, ylab="", ylim=range(-12, 13))
par(new=TRUE)
plot(X, Y_hat, cex=.4, ylab="Y", ylim=range(-12, 13), title("Y ~ 1 + X"), col='red')
# Computing leave-one-out cross validation error
cross_val_1 <- 0
for (i in 1:n){
  X_des_new <- X_des[-i,]
  Y_new <- Y[-i]
  b_hat <- solve(t(X_des_new) %*% X_des_new) %*% t(X_des_new) %*% Y_new
  y_hat <- X_des[i,] %*% b_hat
  #print(cross_val_1)
  print((Y[i] - y_hat)^2)
  cross_val_1 = cross_val_1 + (Y[i] - y_hat)^2
  #print(length(cross_val_1))
}
RSS <- sum((Y - Y_hat)^2)
AIC <- -sum((Y - Y_hat)^2) - length(b_hat)
BIC <- -sum((Y - Y_hat)^2) - length(b_hat)/2*log(n)
Performance[,4] = c(RSS, AIC, BIC, cross_val_1)

# Fitting extended model with cos(X) and sin(X) and X^2
X_des <- cbind(matrix(rep(1, n), n, 1, byrow=TRUE), X, X^2, cos(X), sin(X))
b_hat <- solve(t(X_des) %*% X_des) %*% t(X_des) %*% Y
Y_hat <- X_des %*% b_hat
plot(X, Y, cex=.4, ylab="", ylim=range(-12, 13))
par(new=TRUE)
plot(X, Y_hat, cex=.4, ylab="Y", ylim=range(-12, 13), title("Y ~ 1 + X"), col='red')
# Computing leave-one-out cross validation error
cross_val_1 <- 0
for (i in 1:n){
  X_des_new <- X_des[-i,]
  Y_new <- Y[-i]
  b_hat <- solve(t(X_des_new) %*% X_des_new) %*% t(X_des_new) %*% Y_new
  y_hat <- X_des[i,] %*% b_hat
  #print(cross_val_1)
  print((Y[i] - y_hat)^2)
  cross_val_1 = cross_val_1 + (Y[i] - y_hat)^2
  #print(length(cross_val_1))
}
RSS <- sum((Y - Y_hat)^2)
AIC <- -sum((Y - Y_hat)^2) - length(b_hat)
BIC <- -sum((Y - Y_hat)^2) - length(b_hat)/2*log(n)
Performance[,5] = c(RSS, AIC, BIC, cross_val_1)

### Forward stepwise regression will choose models as follows:
# Y ~ 1
# Y ~ 1 + X
# Y ~ 1 + X + sin(X)
# Y ~ 1 + X + sin(X) + cos(X)
# Stop

### backward stepwise regression will choose models as follows:
# Y ~ 1 + X + X^2 + sin(X) + cos(X) 
# Y ~ 1 + X + sin(X) + cos(X)
# Stop


